sda/cmd/verify/verify.go

```diff
@@ -204,13 +204,20 @@ func main() {
 				continue
 			}
 
-			hr := bytes.NewReader(header)
-			// Feed everything read from the archive file to archiveFileHash
-			mr := io.MultiReader(hr, io.TeeReader(f, archiveFileHash))
+			var c4ghr *streaming.Crypt4GHReader
 
-			c4ghr, err := streaming.NewCrypt4GHReader(mr, *key, nil)
-			if err != nil {
-				log.Errorf("Failed to open c4gh decryptor stream, reson: (%s)", err.Error())
+			for _, key := range archiveKeyList {
+				// Feed data read from the archive file into archiveFileHash
+				mr := io.MultiReader(bytes.NewReader(header), io.TeeReader(f, archiveFileHash))
+
+				c4ghr, err = streaming.NewCrypt4GHReader(mr, *key, nil)
+				if (err == nil) && (c4ghr != nil) {
+					break
+				}
+				log.Warnf("Failed to open C4GH decryptor with key: %v, trying next key. Reason: %s", *key, err.Error())
+			}
```

Comment:
```suggestion
			var key *[32]byte
			for _, key = range archiveKeyList {
				size, err := headers.EncryptedSegmentSize(header, *key)
				if (err == nil) && (size != 0) {
					break
				}
			}

			mr := io.MultiReader(bytes.NewReader(header), io.TeeReader(f, archiveFileHash))
			c4ghr, err := streaming.NewCrypt4GHReader(mr, *key, nil)
			if err != nil {
```

I'm thinking something like this would suffice.
There is not really any point in creating the readers on every iteration in the for loop.